dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/route/RouteService.java

```diff
@@ -0,0 +1,160 @@
+import org.jasypt.encryption.pbe.PBEStringCleanablePasswordEncryptor;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.http.*;
+import org.springframework.http.client.HttpComponentsClientHttpRequestFactory;
+import org.springframework.stereotype.Service;
+import org.springframework.util.StreamUtils;
+import org.springframework.web.client.RestTemplate;
+import org.springframework.web.util.UriComponentsBuilder;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+/**
+ * @author Morten Olav Hansen
+ */
+@Service
+@Slf4j
+@RequiredArgsConstructor
+public class RouteService
+{
+    private final RouteStore routeStore;
+
+    private final ObjectMapper objectMapper;
+
+    @Qualifier( AES_128_STRING_ENCRYPTOR )
+    private final PBEStringCleanablePasswordEncryptor encryptor;
+
+    private static final RestTemplate restTemplate = new RestTemplate();
+
+    static
+    {
+        HttpComponentsClientHttpRequestFactory requestFactory = new HttpComponentsClientHttpRequestFactory();
+        requestFactory.setConnectionRequestTimeout( 1_000 );
+        requestFactory.setConnectTimeout( 5_000 );
+        requestFactory.setReadTimeout( 10_000 );
+        requestFactory.setBufferRequestBody( true );
+
+        restTemplate.setRequestFactory( requestFactory );
+    }
+
+    public Route getDecryptedById( @Nonnull String id )
+    {
+        Route route = routeStore.getByUidNoAcl( id );
+
+        if ( route == null )
+        {
+            return null;
+        }
+
+        try
+        {
+            route = objectMapper.readValue( objectMapper.writeValueAsString( route ), Route.class );
+            decrypt( route );
+        }
+        catch ( JsonProcessingException ex )
+        {
+            log.error( "Unable to create clone of Route with ID " + route.getUid() + ". Please check it's data." );
+            return null;
+        }
+
+        return route;
+    }
+
+    public ResponseEntity<String> exec( Route route, HttpServletRequest request )
+        throws IOException
+    {
+        HttpHeaders headers = new HttpHeaders();
+        route.getHeaders().forEach( headers::add );
+
+        if ( route.getAuth() != null )
+        {
+            route.getAuth().apply( headers );
+        }
+
+        HttpHeaders queryParameters = new HttpHeaders();
+        request.getParameterMap().forEach( ( key, value ) -> queryParameters.addAll( key, Arrays.asList( value ) ) );
+
+        UriComponentsBuilder uriComponentsBuilder = UriComponentsBuilder.fromHttpUrl( route.getUrl() )
+            .queryParams( queryParameters );
+
+        String body = StreamUtils.copyToString( request.getInputStream(), StandardCharsets.UTF_8 );
+
+        HttpEntity<String> entity = new HttpEntity<>( body, headers );
+
+        HttpMethod httpMethod = HttpMethod.resolve( request.getMethod() );
+
+        if ( httpMethod == null )
```

Comment:
Prefer `ObjectUtils.firstNonNull( httpMethod, HttpMethod.GET)` to avoid if statement.

Reply:
Changed to `Objects.requireNonNullElse`